### PR TITLE
fix: edge case crashes with profiled traces

### LINKED ISSUE #3299: Uncaught exception in `profilerForFinishedTracer`
### Platform

iOS

### Installed

Swift Package Manager

### Version

8.13.0

### Steps to Reproduce

Will work on more comprehensive steps to repro async, but for now at least wanted to get this on your radar.

### Expected Result

Ability to call `-[SentryTracer finishWithStatus:]` without triggering internal exceptions. This was working in 8.8.0, but tracing has been broken for us since that release for various reasons, first being a deadlock then with 8.13 we hit this exception.

### Actual Result

> Expected a profiler to be associated with tracer id dfd99f76d40d4af38488db3fff789426.

```
#0  __exceptionPreprocess ()
#1  objc_exception_throw ()
#2  -[NSAssertionHandler handleFailureInFunction:file:lineNumber:description:] ()
#3  profilerForFinishedTracer at /<redacted>/sentry-cocoa/Sources/Sentry/Profiling/SentryProfiledTracerConcurrency.mm:118
#4  +[SentryProfiler collectProfileBetween:and:forTrace:onHub:] at /<redacted>/sentry-cocoa/Sources/Sentry/SentryProfiler.mm:394
#5  +[SentryProfiler createProfilingEnvelopeItemForTransaction:] at /<redacted>/sentry-cocoa/Sources/Sentry/SentryProfiler.mm:363
#6  -[SentryTracer captureTransactionWithProfile:] at /<redacted>/sentry-cocoa/Sources/Sentry/SentryTracer.m:540
#7  -[SentryTracer finishInternal] at /<redacted>/sentry-cocoa/Sources/Sentry/SentryTracer.m:528
#8  -[SentryTracer canBeFinished] at /<redacted>/sentry-cocoa/Sources/Sentry/SentryTracer.m:428
#9  -[SentryTracer finishWithStatus:] at /<redacted>/sentry-cocoa/Sources/Sentry/SentryTracer.m:400
```

### Are you willing to submit a PR?

Yes, if I can figure out a fix

### PR DESCRIPTION
**Pull request by armcknight**
fixes #3299

Thanks for @philipphofmann finding some [cases](https://github.com/getsentry/sentry-cocoa/issues/3299#issuecomment-1787556279) where this could potentially happen, we'll patch up those cases until we hear any other information from the customer that reported the original issue.

This introduces an immutable internal tracer ID for concurrency tracking purposes, and avoids querying for profiling data if the profiler failed initializing.

#skip-changelog

### CODE REVIEW

File: Sources/Sentry/include/SentryTracer+Private.h [hunk 1]
```diff
@@ -5,4 +5,10 @@ SentryTracer ()
 
 @property (nonatomic, strong) SentryHub *hub;
 
+/**
+ * We need an unchanging identifier to track concurrent tracers against a static profiler.
+ * @c SentryTracer.traceId can be changed by consumers so is unfit for this purpose.
+ */
+@property (nonatomic, strong) SentryId *concurrencyID;
```
Comment (philipphofmann) [inline Sources/Sentry/include/SentryTracer+Private.h:12]: `h`: This has to be immutable.
```suggestion
@property (nonatomic, readonly, strong) SentryId *concurrencyID;
```

Comment (philipphofmann) [inline Sources/Sentry/include/SentryTracer+Private.h:12]: `l`: I think we could use the ID also for other purposes than concurrency. We could also call it something like `immutableID`, `immutableTraceID`, etc.

File: Sources/Sentry/include/SentryTracer+Private.h [hunk 2]
```diff
@@ -5,4 +5,10 @@ SentryTracer ()
 
 @property (nonatomic, strong) SentryHub *hub;
 
+/**
+ * We need an unchanging identifier to track concurrent tracers against a static profiler.
```
Comment (philipphofmann) [inline Sources/Sentry/include/SentryTracer+Private.h:9]: Thanks for this comment 👏

File: Sources/Sentry/include/SentryProfiler.h [hunk 3]
```diff
@@ -44,7 +44,7 @@ SENTRY_EXTERN_C_END
 /**
  * Start a profiler, if one isn't already running.
  */
-+ (void)startWithTracer:(SentryId *)traceId;
++ (BOOL)startWithTracer:(SentryId *)traceId;
```
Comment (philipphofmann) [inline Sources/Sentry/include/SentryProfiler.h:47]: `m`: I think we should rename `traceID` to `concurrencyID` to state it's not the traceID. I would also rename all occurrences in `traceID` in `SentryProfiledTracerConcurrency`.

File: Sources/Sentry/SentryProfiler.mm [hunk 4]
```diff
@@ -331,16 +331,17 @@ + (void)startWithTracer:(SentryId *)traceId
         trackProfilerForTracer(_gCurrentProfiler, traceId);
         // record a new metric sample for every concurrent span start
         [_gCurrentProfiler->_metricProfiler recordMetrics];
-        return;
+        return NO;
```
Comment (philipphofmann) [inline Sources/Sentry/SentryProfiler.mm:334]: `m`: Shouldn't this be `YES`? We are profiling in that case.

```suggestion
        return YES;
```

**Review by philipphofmann — COMMENTED** [review]
I think we should add a changelog entry for this, as this PR solves a problem.

File: Sources/Sentry/include/SentryTracer+Private.h (hunk 1 above)
Comment (armcknight) [reply Sources/Sentry/include/SentryTracer+Private.h:12]: I tried this and redeclaring it as readwrite in SentryTracer.m and it failed with error: duplicate declaration

Comment (armcknight) [reply Sources/Sentry/include/SentryTracer+Private.h:12]: i'll call it `internalID`

File: Sources/Sentry/include/SentryProfiler.h (hunk 3 above)
Comment (armcknight) [reply Sources/Sentry/include/SentryProfiler.h:47]: i'll call it `internalID` to match the declaration's name in SentryTracer

File: Sources/Sentry/SentryProfiler.mm (hunk 4 above)
Comment (armcknight) [reply Sources/Sentry/SentryProfiler.mm:334]: ay, good catch!

File: Sources/Sentry/include/SentryTracer+Private.h (hunk 1 above)
Comment (armcknight) [reply Sources/Sentry/include/SentryTracer+Private.h:12]: nm, don't even need the readwrite redeclare, just have to set the ivar in init

**Review by indragiek — APPROVED** [review]

**Comment by philipphofmann** [comment]
Thanks for the quick fix @armcknight 👏

**Pull request merged**